### ISSUE TITLE
Concurrency issues in cache

### ISSUE THREAD

**Issue by mrent** [filipw/AspNetWebApi-OutputCache#113]
Following scenario causes invalid cache entries: a user requests data, a cache miss occurs and the data is retrieved. Before this data is added to the cache, a second request is received and processed that invalidates the data retrieved in the first request (goes all the way through, including cache operations). After this the retrieved data for the first the request is added to the cache.

Technically you need to know if data was potentially invalidated since the point in time you started retrieval rather than since the point in time you stored the data in the cache.

Typically you would not see consequences of this issue unless you are running a site under high load.

**Comment by jods4** [comment]
I noticed this today and was about to fill a bug about it. 
To illustrate, if you a read method that is cached:
`[Cached] return x;`
and a write that invalidates the cache:
`x++; cache.Remove()`.
Things may happen in this order if you're unlucky:
```
// Thread 1 calls read
return x;
// Thread 2 calls write
x++;
cache.Remove();
// Thread 1
cache.Add(x);
```
And you have an obsolete value in the cache.

I think the proper way to fix this is to create the `baseKey` before the action executes (i.e. inside `OnActionExecuting` rather than `OnActionExecuted`) and create a `CacheDependency` at that time. Then during `OnActionExecuted` you can add you cache entry with its dependency. If it was invalidated meanwhile, the cache will immediately discard it.

Note that this fix requires changing the `IApiOutputCache` interface, because you cannot just pass a `string` for the dependant cache entry, you need some kind of `CacheDependency` interface.

BTW: is this project still maintained? I see lots of bugs reported but not a single commit, or even a comment?

**Comment by mrent** [comment]
yes, a proper way (which is how I handled it also) is creating a cache dependency during the request and discard it when invalidated. 

Not sure if this is maintained. It is one of the few Web API cache providers out there right now, but it really does need some care. Maybe can replace this once vNext is out.
